shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala

```diff
@@ -350,61 +366,84 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
   val BYTES_ALLOWED_PER_BATCH: Long = _2GB - APPROX_PAR_META_DATA
 
   private[rapids] def compressColumnarBatchWithParquet(
-      gpuCB: ColumnarBatch,
+      oldGpuCB: ColumnarBatch,
       schema: StructType): List[ParquetCachedBatch] = {
-    val estimatedRowSize = scala.Range(0, gpuCB.numCols()).map { idx =>
-      gpuCB.column(idx).asInstanceOf[GpuColumnVector].getBase.getDeviceMemorySize / gpuCB.numRows()
+    val estimatedRowSize = scala.Range(0, oldGpuCB.numCols()).map { idx =>
+      oldGpuCB.column(idx).asInstanceOf[GpuColumnVector]
+          .getBase.getDeviceMemorySize / oldGpuCB.numRows()
     }.sum
-    val rowsInBatch = (BYTES_ALLOWED_PER_BATCH / estimatedRowSize).toInt
-    val splitIndices = scala.Range(rowsInBatch, gpuCB.numRows(), rowsInBatch)
-    val buffers = new ListBuffer[ParquetCachedBatch]
-    if (splitIndices.nonEmpty) {
-      val splitVectors = new ListBuffer[Array[ColumnVector]]
-      try {
-        for (index <- 0 until gpuCB.numCols()) {
-          splitVectors +=
-              gpuCB.column(index).asInstanceOf[GpuColumnVector].getBase.split(splitIndices: _*)
+    // We want to map the name of the cols we are casting so we can cast them back when reading
+    // back the cache
+    val castMap = new scala.collection.mutable.HashMap[String, DType]()
```

Comment:
This data structure fundamentally can never work for nested types.  You are storing a single DType per column.  What happens if I have a struct with multiple nested Decimal values in it?  What happens if I also have a mixture of LongType and IntTypes in there too?

Reply:
Ugh, I meant to use Map[String, ArrayBuffer[DType]]. I have simplified the logic to not need a map. PTAL

Reply:
Why is this needed at all??  You have the schema of the output data.  You know how you would transform it and it should be consistent. So you should just need to walk the output schema along with the data structure returned from the parquet reader and bit cast it when you see something different.  Or throw an exception if the types are so different that they cannot be bit cast.

Reply:
Yeah, that's exactly what I have done in the latest commit. 

I still  need to work on renaming all the child columns in a nested structure, otherwise the parquet reader barfs